architecture/datapath/http-filter-chain.md

```diff
@@ -0,0 +1,97 @@
+
+### Routing phase
+
+Istio configures routing using xDS routing configuration and does not rely on additional HTTP filters to determine the route and the upstream cluster.
+
+### Application security
+
+Istio defines two steps within this phase:
+
+* _Authentication_: extracting the secure attributes from the request to influence policy decision and route selection.
+* _Authorization_: enforcement of the policies.
+
+Authentication should precede authorization so that the policy takes into account the full set of attributes, and also because Istio allows authentication filters to affect routing.
+For example, JWT authentication filters can influence the routing decision via a [claim-based routing feature](https://istio.io/latest/docs/tasks/security/authentication/jwt-route/).
+
+A special consideration should be applied to the CORS policy filter. This filter short-circuits the preflight requests by explicitly allowing them. Because browsers are [not supplying the additional authentication headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#examples_of_access_control_scenarios), placing CORS after JWT authentication causes unexpected denial of the preflight requests. Therefore,
+the correct placement for CORS is to precede JWT and other authentication filters.
+
+### Traffic management
+
+In Istio, traffic management features include the fault injection, grpc-web transcoding, persistent session load balancing, and mTLS ALPN overriding. These features should be logically placed next in the chain, after the application security filters.
+
+### Telemetry
+
+Telemetry filters are expected to observe the traffic and typically have no visible effect on the data path. Therefore, these filters in practice can be placed anywhere in the chain, but should be placed last to capture the final request form.
+
+## Istio filter order specification
+
+If we follow the principles outlined above, the following filter chain order should take into account the entirety of Istio HTTP filter extensions:
+
+| Filter | Phase | Purpose |
+| --- | --- | --- |
+| `metadata_exchange` | Application Security / Authentication | Peer metadata attributes (currently insecure) |
+| `wasm`              | Application Security / Authentication | Wasm extension per [AUTHN phase](https://istio.io/latest/docs/reference/config/proxy_extensions/wasm-plugin/#PluginPhase) |
+| `cors`              | Application Security / Authentication | Allow CORS requests per CORS policy |
+| `jwt_authn`         | Application Security / Authentication | JWT attributes and routing |
+| `wasm`              | Application Security / Authorization | Wasm extension per [AUTHZ phase](https://istio.io/latest/docs/reference/config/proxy_extensions/wasm-plugin/#PluginPhase) |
+| `ext_authz`         | Application Security / Authorization | Istio CUSTOM authorization |
+| `rbac`              | Application Security / Authorization | Istio local authorization |
```

Comment:
put rbac before ext authz, local first

Reply:
No, the order is right per

```

Reply:
I put up this because in k8s admission controllers, the order are well designed. 
The mutating/validating webhooks are put last.

```
var AllOrderedPlugins = []string{
	admit.PluginName,                        // AlwaysAdmit
	autoprovision.PluginName,                // NamespaceAutoProvision
	lifecycle.PluginName,                    // NamespaceLifecycle
	exists.PluginName,                       // NamespaceExists
	scdeny.PluginName,                       // SecurityContextDeny
	antiaffinity.PluginName,                 // LimitPodHardAntiAffinityTopology
	limitranger.PluginName,                  // LimitRanger
	serviceaccount.PluginName,               // ServiceAccount
	noderestriction.PluginName,              // NodeRestriction
	nodetaint.PluginName,                    // TaintNodesByCondition
	alwayspullimages.PluginName,             // AlwaysPullImages
	imagepolicy.PluginName,                  // ImagePolicyWebhook
	podsecurity.PluginName,                  // PodSecurity
	podnodeselector.PluginName,              // PodNodeSelector
	podpriority.PluginName,                  // Priority
	defaulttolerationseconds.PluginName,     // DefaultTolerationSeconds
	podtolerationrestriction.PluginName,     // PodTolerationRestriction
	eventratelimit.PluginName,               // EventRateLimit
	extendedresourcetoleration.PluginName,   // ExtendedResourceToleration
	label.PluginName,                        // PersistentVolumeLabel
	setdefault.PluginName,                   // DefaultStorageClass
	storageobjectinuseprotection.PluginName, // StorageObjectInUseProtection
	gc.PluginName,                           // OwnerReferencesPermissionEnforcement
	resize.PluginName,                       // PersistentVolumeClaimResize
	runtimeclass.PluginName,                 // RuntimeClass
	certapproval.PluginName,                 // CertificateApproval
	certsigning.PluginName,                  // CertificateSigning
	certsubjectrestriction.PluginName,       // CertificateSubjectRestriction
	defaultingressclass.PluginName,          // DefaultIngressClass
	denyserviceexternalips.PluginName,       // DenyServiceExternalIPs

	// new admission plugins should generally be inserted above here
	// webhook, resourcequota, and deny plugins must go at the end

	mutatingwebhook.PluginName,           // MutatingAdmissionWebhook
	validatingadmissionpolicy.PluginName, // ValidatingAdmissionPolicy
	validatingwebhook.PluginName,         // ValidatingAdmissionWebhook
	resourcequota.PluginName,             // ResourceQuota
	deny.PluginName,                      // AlwaysDeny
}
```